registry/storage/schema2manifesthandler.go

```diff
@@ -79,8 +82,22 @@ func (ms *schema2ManifestHandler) verifyManifest(ctx context.Context, mnfst sche
 			errs = append(errs, distribution.ErrManifestBlobUnknown{Digest: target.Digest})
 		}
 
-		for _, fsLayer := range mnfst.References() {
-			_, err := ms.repository.Blobs(ctx).Stat(ctx, fsLayer.Digest)
+		for _, fsLayer := range mnfst.Layers {
+			var err error
+			if fsLayer.MediaType != schema2.MediaTypeForeignLayer {
+				if fsLayer.ForeignURL == "" {
+					_, err = ms.repository.Blobs(ctx).Stat(ctx, fsLayer.Digest)
+				} else {
+					err = errors.New("unexpected foreign URL on layer")
+				}
+			} else {
+				// Clients download this layer from an external URL, so do not check for
+				// its presense.
+				u, err := url.ParseRequestURI(fsLayer.ForeignURL)
```

Comment:
Well, Parse allows e.g. `http://foo/bar/#baz`, which seems undownloadable. ParseRequestURI seems to disallow the `#baz`.